jax_test3.py

```diff
@@ -55,65 +55,51 @@ def E_NN(N, Zs, *Rs):
     return e_nn
 
 
-def get_grad(coordinates, charges):
-    gradient = []
-    natoms = len(charges)
-    state = np.zeros(
-        2 + 3 * natoms, dtype=int
-    )  # makes an array that derv can work with
-    flat_coords = coordinates.reshape(3 * natoms)
-    for item in range(2, len(state)):
-        state[item] = 1  # change the array to the wanted derivation for the hessian
-        eval_E = derv(
-            E_NN,
-            [charges.size, charges, *flat_coords],
-            state,
-        )
-        gradient.append(eval_E)
-        state[item] = 0  # reset array
-
-    gradient = np.array(gradient)
-    gradient = gradient.reshape((natoms, 3))
-    return gradient
-
-
-# this will generate the full hessian
-def get_hessian(coordinates, charges):
-    natoms = len(charges)
-    state = np.zeros(2 + 3 * natoms, dtype=int)
-    hessian = np.zeros(
-        (natoms, 3) * 2
-    )  # reshaping the hessian gave errors so here's created a priori
-    # these will keep track whether we are on the x y or z coordinate
-    i_xyz = 0
-    a_xyz = 0
-    flat_coords = coordinates.reshape(3 * natoms)
-    for i in range(2, len(state)):
-        state[i] += 1  # change the item state
-        for a in range(2, len(state)):
-            state[a] += 1
+def derv_plus(current, stop, state, dervs, charges, flat_coords):
+    if current != stop:
+        current += 1
+        for i in range(2, len(state)):
+            state[i] += 1  # change the item state
+
+            derv_plus(current, stop, state, dervs, charges, flat_coords)
+
+            state[i] -= 1
+    else:
+        for i in range(2, len(state)):
+            state[i] += 1  # change the item state
+
             eval_E = derv(
                 E_NN,
                 [charges.size, charges, *flat_coords],
                 state,
             )
-            # print(f"mapping to {(i-2)//3},{i_xyz},{(a-2)//3},{a_xyz}") METADATA
-            hessian[(i - 2) // 3, i_xyz, (a - 2) // 3, a_xyz] = eval_E
-            state[a] -= 1  # reset
-            a_xyz += 1  # change x y z
-            a_xyz *= a_xyz != 3  # reset if more than 2 (z)
+            dervs.append(eval_E)
+
+            state[i] -= 1
+
+
+# this will generate the derivatives layer
+def get_derv(coordinates, charges, layer):
```

Comment:
This function is now specific for the computation of derivatives of the E_NN function. I think it's better to give a more specific name:
```suggestion
def E_NN_derivatives(coordinates, charges, order):
```